grblas/matrix.py

```diff
@@ -777,21 +813,85 @@ def _prep_for_assign(self, resolved_indexes, value):
+                        expr_repr = "[{1}, [{3} cols]] = {0.name}"
+                    # SS, SuiteSparse-specific: assume efficient vector with single scalar
+                    value_vector << value
+
+                    # Row-only selection
+                    row_index = rows
+                    delayed = MatrixExpression(
+                        method_name,
+                        cfunc_name,
+                        [value_vector, row_index, cols, colsize],
+                        expr_repr=expr_repr,
+                        nrows=self._nrows,
+                        ncols=self._ncols,
+                        dtype=self.dtype,
+                    )
+                elif colsize is None and rowsize is not None:
+                    if is_submask:
+                        # SS, SuiteSparse-specific: subassign
+                        cfunc_name = "GrB_Col_subassign"
+                        value_vector = Vector.new(value.dtype, size=mask.mask.size, name="v_temp")
+                    else:
+                        cfunc_name = "GrB_Col_assign"
+                        value_vector = Vector.new(value.dtype, size=rowsize, name="v_temp")
+                    # SS, SuiteSparse-specific: assume efficient vector with single scalar
+                    value_vector << value
+
+                    # Column-only selection
+                    col_index = cols
+                    delayed = MatrixExpression(
+                        method_name,
+                        cfunc_name,
+                        [value_vector, rows, rowsize, col_index],
+                        expr_repr="[[{2} rows], {3}] = {0.name}",
+                        nrows=self._nrows,
+                        ncols=self._ncols,
+                        dtype=self.dtype,
+                    )
+                elif colsize is None and rowsize is None:
+                    # Matrix object, Vector mask, scalar index
+                    raise TypeError(
+                        "Unable to use Vector mask on single element assignment to a Matrix"
+                    )
+                else:
+                    # Matrix object, Vector mask, Matrix index
+                    raise TypeError("Unable to use Vector mask on Matrix assignment to a Matrix")
+            else:
+                if rowsize is None:
+                    rows = _CArray([rows.scalar.value])
+                    rowsize = _CScalar(1)
+                if colsize is None:
+                    cols = _CArray([cols.scalar.value])
+                    colsize = _CScalar(1)
+                if is_submask:
+                    # SS, SuiteSparse-specific: subassign
+                    cfunc_name = f"GrB_Matrix_subassign_{value.dtype}"
+                    expr_repr = "[[{2} rows], [{4} cols]](%s) = {0}" % mask.name
+                else:
+                    cfunc_name = f"GrB_Matrix_assign_{value.dtype}"
+                    expr_repr = "[[{2} rows], [{4} cols]] = {0}"
```

Comment:
How we handle scalar indices here results in some interesting behavior.  The code in this else statement handles scalar values on the RHS and Matrix or no mask.  Specifically:
```python
C[I, J](M) << c
C[I, j](M) << c
C[i, J](M) << c
C[i, j](M) << c
C[I, J] << c
C[I, j](<< c
C[i, J] << c
C[i, j] << c
C(M)[I, J] << c
C(M)[I, j](<< c
C(M)[i, J] << c
C(M)[i, j] << c
```
I don't think we should allow the following:
```python
C[I, j](M) << c
C[i, J](M) << c
C[i, j](M) << c
```
For assign, `C(mask)[indices] << value`, the mask must either match the matrix `C` or the indices (such as being a vector).

For subassign `C[indices](mask) << value`, the mask must match the indices.  There is no other option.

For both assign and subassign, the value must either match indices or be a scalar.

We do not currently treat a matrix with a dimension of length 1 as a vector, nor do we ever treat a vector as a matrix with a dimension of length 1.  Both of these behaviors are, of course, possible and safe, but are clearly "not yet".  Similarly, we could do the same for vector or matrix with one element and scalars.

Reply:
I agree with not allowing those options. Additionally, I wonder if we should disallow:
```python
C(M)[i, j] << c
```
Masks for a single element feel strange. But I'm not 100% against it either.

Reply:
It's cheap and low-complexity to support `C(M)[i, j] << c`, so I don't mind supporting mask (and accum, etc) here with a single element update with a scalar.

But, this does raise the option of `C[i, j].new(input_mask=Mask)`, which is probably fine-ish.  The assignment-equivalent of this is `c(input_mask=Mask) << C[i, j]`.  It's a little weird to use `input_mask` like this in a Scalar, but consistent.  `C[I, j].new(input_mask=Mask)` with vector result doesn't feel weird to me.

btw, partial update has been pushed (no `input_mask` yet).  There are a lot of error messages that need written to help guide people.